### ISSUE TITLE
Wrong type inference for method of type

### ISSUE THREAD

**Issue by amotzop** [microsoft/pyright#3361]
**Describe the bug**
There is some bug in the return type of the `type.__subclasses__` method (when working with a TypeVar bound to `type`).
It doesn't seem like a `typeshed` problem because mypy behaves differently.

**To Reproduce**
pyright will throw an error on the code below (`Expression of type "list[Type[object]]" cannot be assigned to return type "list[Type[Cls@func]]"`)
This also happens if we define `Cls` as bound to `type` and replace `type[Cls]` with `Cls` in the function annotations

**Expected behavior**
The return type of `__subclasses__` should be as annotated.

**Screenshots or Code**
```python
from typing import TypeVar

Cls = TypeVar("Cls")

def func(cls: type[Cls]) -> list[type[Cls]]:
    return cls.__subclasses__()
```

**VS Code extension or command-line**
This happens both in VS Code and in the command-line tool (version 1.1.239)

**Comment by erictraut** [comment]
Thanks for the bug report. Yeah, it looks like the type is being evaluated incorrectly here. I'll investigate further.

**Comment by erictraut** [comment]
This was broken by some special-case logic that I added a while back to address another bug report that involved a somewhat dodgy piece of code. I've undone that special case handling because it's arguably incorrect, and it causes problems like the one involved in this bug report.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.240, which I just published. It will also be included in the next release of pylance.
